### PR TITLE
Refactor repo and added metadata endpoints 

### PR DESCRIPTION
**Pull request by alfrunes**
I started refactoring this repository to comply with the source tree of
other microservices. Then I got a bit carried away and added a couple of
new changes and endpoints:

Changes:
 - Moved workflow map to the datastore and a workflow collection in
   mongo (the map is treated as a cache, and the collection provides
   persistency).
 - Fixed a bug where the tailing cursor to the mongodb message bus would
   instantly die due to empty collection.
 - Made the status enumerated int type as updating a document that
   imposes change in size in a capped collection is prohibited.
 - Added status filtering to the job scheduler routine such that the
   worker doesn't reschedule all old jobs on restart.
 - Removed JobStatus and merged the struct into Job.
 - Modified json tags to comply more with conductor workflow
   definitions.

Endpoints:
 - GET  /api/metadata/workflows > Lists all workflows (JSON-list)
 - POST /api/metadata/workflows > Add new workflow

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

### CODE REVIEW

File: api/http/workflow_test.go [hunk 1]
```diff
@@ -48,60 +50,66 @@ func TestWorkflowFoundButMissingParameters(t *testing.T) {
...
+//	payload := `{
+//      "key": "value"
+//	}`
+//
+//	req, _ := http.NewRequest("POST", "/api/workflow/test", strings.NewReader(payload))
+//	router.ServeHTTP(w, req)
+//
+//	assert.Equal(t, 200, w.Code)
+//	assert.Equal(t, 1, len(dataStore.Jobs))
+//}
```
Comment (tranchitella) [inline api/http/workflow_test.go]: Why is this tested commented out? Oversight?

File: config/config.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ const (
 	// SettingListen is the config key for the listen address
 	SettingListen = "listen"
 	// SettingListenDefault is the default value for the listen address
-	SettingListenDefault = ":8080"
+	SettingListenDefault = "workflows:8080"
```
Comment (tranchitella) [inline config/config.go]: Why?

File: api/http/workflow_test.go (hunk 1 above)
Comment (alfrunes) [reply api/http/workflow_test.go]: Yeah, I forgot about this one.

File: config/config.go (hunk 2 above)
Comment (alfrunes) [reply config/config.go]: Oops.. This is just a goof from when I was playing around. I didn't actually notice as the default `config.yaml` file overwrites it.

**Comment by alfrunes** [comment]
@tranchitella Could you please take a second look at some of the recent changes? (You can ignore the vendor and ci commit).

@lluiscampos Could you take a look at the CI script?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,90 @@
+image: golang:1.11
+
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: We need to delete the existing `.gitlab-ci.yaml` file

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  script:
+    # Install cyclomatic dependency analysis tool
+    - go get -u github.com/fzipp/gocyclo
+    - curl -sL https://deb.nodesource.com/setup_11.x | bash -
+    - apt-get -qq update
+    - apt-get install -qy --allow-unauthenticated python3-pip nodejs liblzma-dev
+    - pip3 install pyyaml
+    - npm install -g swagger-cli
+    # Get our own Swagger verifier
+    - wget https://raw.githubusercontent.com/mendersoftware/autodocs/master/verify_docs.py
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: This repo does not have API docs. This is unnecessary

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    # fail if any have been modified.
+    - if [ -n "$(go fmt)" ]; then echo 'Code is not formatted with "go fmt"'; false; fi
+    # Perform static code analysys
+    - go vet `go list ./... | grep -v vendor`
+    # Fail builds when the cyclomatic complexity reaches 20 or more
+    - gocyclo -over 20 `find . -iname '*.go' | grep -v 'vendor' | grep -v '_test.go'`
+    # Verify that the Swagger docs are valid
+    - swagger-cli validate docs/*.yml
+    # Verify that the Swagger docs follow the autodeployment requirements
+    - python3 verify_docs.py `find docs -name "*.yml"`
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: Same here.

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    - /go/src/golang.org
+    - /go/src/gopkg.in
+    - node_modules
+
+variables:
+  DOCKER_HOST: tcp://docker:2375/
+  DOCKER_DRIVER: overlay2
+
+services:
+  - docker:dind
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: We don't need `docker:dind` service

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  - mongo
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: Unsure about this one: is it needed for unit tests?

I don't think so as it seems that we are installing mongo the old way before unit tests

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    - git branch -m temp-branch || true
+    # Git trick: Fetch directly into our local branches instead of remote
+    # branches.
+    - git fetch origin 'refs/heads/*:refs/heads/*'
+    # Get last remaining tags, if any.
+    - git fetch --tags origin
+
+    - apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 2930ADAE8CAF5059EE73BB4B58712A2291FA4AD5
+    - echo "deb http://repo.mongodb.org/apt/debian stretch/mongodb-org/3.6 main" | tee /etc/apt/sources.list.d/mongodb-org-3.6.list
+    - apt-get -qq update && apt-get install -qy --allow-unauthenticated mongodb-org-server=3.6.11  liblzma-dev
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: Is this necessary at all?

File: .gitlab-ci.yml [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    - cp -r $CI_PROJECT_DIR /go/src/github.com/mendersoftware/workflows
+    - ln -s /go/src/github.com/mendersoftware /go/src/_/builds/mendersoftware
+    - cd /go/src/github.com/mendersoftware/workflows
+    - go list ./... | grep -v vendor | xargs -n1 -I {} -P 4 go test -v -covermode=atomic -coverprofile=../../../{}/coverage.txt {} || exit $? ;
+    - mkdir -p tests/unit-coverage && find . -name 'coverage.txt' -exec cp --parents {} ./tests/unit-coverage \;
+    - tar -cvf $CI_PROJECT_DIR/unit-coverage.tar tests/unit-coverage
+  artifacts:
+    expire_in: 2w
+    paths:
+      - unit-coverage.tar
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: Missing:
- Commits and license checks: these can be done vendoring the mendertesting module and adding the unit tests
- Publishing code coverage to codecov (probably also setting up codecov)

Comment (lluiscampos) [reply .gitlab-ci.yml]: I can to these myself in a separate PR if you want.

File: .gitlab-ci.yml [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  paths:
+    - /apt-cache
+    - /go/src/github.com
+    - /go/src/golang.org
+    - /go/src/gopkg.in
+    - node_modules
+
+variables:
+  DOCKER_HOST: tcp://docker:2375/
+  DOCKER_DRIVER: overlay2
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: Unnecessary

File: .gitlab-ci.yml [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+before_script:
+  - export DOCKER_REPOSITORY="mendersoftware/workflows"
+  - export DOCKER_TAG=${CI_COMMIT_REF_SLUG:-master}
+  - export SERVICE_IMAGE=$DOCKER_REPOSITORY:$DOCKER_TAG
+  - export COMMIT_TAG="$CI_COMMIT_REF_SLUG"_"$CI_COMMIT_SHA"
+
+test:static:
+  stage: test
+  dependencies: []
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: Here and below: you don't need an empty dependencies list, better to just omit it.

File: .gitlab-ci.yml [hunk 12]
```diff
@@ -0,0 +1,90 @@
+image: golang:1.11
```
Comment (lluiscampos) [inline .gitlab-ci.yml]: Should we use a newer golang for this service? As it is new, I would suggest trying with latest golang 1.13

File: .gitlab-ci.yml (hunk 9 above)
Comment (lluiscampos) [reply .gitlab-ci.yml]: Missing also:
- Publish into docker registry (private one)

Comment (lluiscampos) [reply .gitlab-ci.yml]: Missing also:
- Docker build!

**Comment by tranchitella** [comment]
@alfrunes looks okay to me!

File: .gitlab-ci.yml (hunk 4 above)
Comment (alfrunes) [reply .gitlab-ci.yml]: Yes it does :)

File: .gitlab-ci.yml (hunk 7 above)
Comment (alfrunes) [reply .gitlab-ci.yml]: Nah, we don't need it yet...

File: .gitlab-ci.yml (hunk 9 above)
Comment (alfrunes) [reply .gitlab-ci.yml]: Could you add this to a separate PR @lluiscampos ?

Comment (alfrunes) [reply .gitlab-ci.yml]: BTW: the repo misses a Dockerfile

File: .gitlab-ci.yml (hunk 3 above)
Comment (alfrunes) [reply .gitlab-ci.yml]: Should be taken care of now, I rebased on top of master and deleted it.

File: .gitlab-ci.yml (hunk 9 above)
Comment (lluiscampos) [reply .gitlab-ci.yml]: Ok. Please assign to me the JIRA ticket after this merge.

**Pull request merged**